go-controller/pkg/crd/egressfirewall/v1/types.go

```diff
@@ -0,0 +1,86 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EgressNetworkFirewallRuleType indicates whether an EgressNetworkFirewallRule allows or denies traffic
+// +kubebuilder:validation:Pattern=^Allow|Deny$
+type EgressFirewallRuleType string
+
+const (
+	EgressFirewallRuleAllow EgressFirewallRuleType = "Allow"
+	EgressFirewallRuleDeny  EgressFirewallRuleType = "Deny"
+)
+
+// +genclient
+// +resource:path=egressfirewall
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:printcolumn:name="EgressFirewall Status",type=string,JSONPath=".status.egressFirewallStatus"
+// EgressFirewall describes the current egress firewall for a Namespace.
+// Traffic from a pod to an IP address outside the cluster will be checked against
+// each EgressFirewallRule in the pod's namespace's EgressFirewall, in
+// order. If no rule matches (or no EgressFirewall is present) then the traffic
+// will be allowed by default.
+type EgressFirewall struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired behavior of EgressFirewall.
+	Spec EgressFirewallSpec `json:"spec"`
+	// Observed status of EgressFirewall
+	// +optional
+	Status EgressFirewallStatus `json:"status,omitempty"`
+}
+
+// EgressFirewallStatus is the status of the EgressFirewall Object
+type EgressFirewallStatus struct {
+	// egressFirewallStatus contains the status of the EgressFirewall object
+	EgressFirewallObjectStatus string `json:"egressFirewallStatus"`
+}
+
+// EgressFirewallSpec is a desired state description of EgressFirewall.
+type EgressFirewallSpec struct {
+	// a collection of egress firewall rule objects
+	Rules []EgressFirewallRule `json:"egress"`
+}
+
+// EgressFirewallRule is a single egressfirewall rule object
+type EgressFirewallRule struct {
+	// type marks this as an "Allow" or "Deny" rule
+	Type EgressFirewallRuleType `json:"type"`
+	// ports specify what ports and protocols the rule applies to
+	// +optional
+	Ports []EgressFirewallPort `json:"ports,omitempty"`
+	// to is one or more targets that traffic is allowed/denied to
+	To EgressFirewallDestination `json:"to"`
+}
+
+// EgressFirewallPort specifies the port to allow or deny traffic to
+type EgressFirewallPort struct {
+	// protocol (tcp, udp, sctp) that the traffic must match.
+	// +kubebuilder:validation:Pattern=^tcp|udp|sctp$
+	Protocol string `json:"protocol"`
+	// port that the traffic must match
+	// +optional
+	// +kubebuilder:validation:Minimum:=1
+	// +kubebuilder:validation:Maximum:=65535
+	PortNum int32 `json:"port,omitempty"`
+}
+
+// EgressFirewallDestination is the endpoint that traffic is either allowed or denied, can be either a dnsName or a cidrSelector
+type EgressFirewallDestination struct {
+	// cidrSelector is the CIDR range to allow/deny traffic to. If this is set, dnsName must be unset.
+	CIDRSelector string `json:"cidrSelector,omitempty"`
```

Comment:
This has an `omitempty` tag but no `+optional`

Should this field be optional by the way, or the entire `EgressFirewallDestination`?

Reply:
it will become optional when the dns is added